### PR TITLE
feat: Support GRACE PERIOD syntax on WITHIN clause with simple window

### PR DESCRIPTION
**Pull request by spena**
### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_
Part of [KLIP-36](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-36-grace-stream-stream-joins.md), this PR adds support of the `GRACE PERIOD` syntax in the `WITHIN` clause for joins with simple windows. 

The new syntax is optional and it is used like this:
```
SELECT * FROM s1 JOIN s2 WITHIN 1 MINUTE GRACE PERIOD 30 SECONDS ON <condition>
SELECT * FROM s1 JOIN s2 WITHIN (1 MINUTE, 2 MINUTE) GRACE PERIOD 30 SECONDS ON <condition>
```

The PR also updates the `QueryAnonymizer` and its tests.

Follow-up PRs will include the following:
- Call the grace() API in kstreams
- QTT tests for joins

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._
Updated unit tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4
Comment (mjsax) [inline ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:231]: Where is the `BEFORE` and `AFTER` keyword?

Comment (mjsax) [inline ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:231]: Do we need a 4th line for `before/after with grace` ?

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java
Comment (mjsax) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java:915]: Should we rename this method to `anonymizeSizeAndUnitFromJoinWindowSize` (also, should we remove `AndUnit` from the name, as we preserve the units? Or is the method incorrect and it _should_ indeed anonymize the unit, too)?

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java
Comment (mjsax) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java:905]: Should we unify this method with the existing `getSizeAndUnitFromJoinWindowSize` ?

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4
Comment (spena) [reply ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:231]: I'll add it in a follow-up PR (part of parallelized work).

Comment (spena) [reply ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:231]: I'll add it in a follow-up PR (part of parallelized work).

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java
Comment (spena) [reply ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java]: I just renamed to `anonymizeJoinWindowSize`. The anonymizer is not changing the unit so I removed it.

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java
Comment (spena) [reply ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java]: I made some changes on this method.

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java
Comment (mjsax) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java:899]: Just realizing that this error message does not match our docs:
```
 [ WITHIN [(before TIMEUNIT, after TIMEUNIT) | N TIMEUNIT] ]
```

But the error message aligns to the grammar definition. Something seems to be off?

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java
Comment (mjsax) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java:600]: As above. Seems `AFTER` and `BEFORE` keywords are missing?

**Comment by mjsax** [comment]
Should we extend some existing QTT tests, to set a grace period and to test that "late" input is dropped for inner/left/outer stream-stream joins? 

@vcrfxia also just adds new n-way joins test that we might want to extend (https://github.com/confluentinc/ksql/pull/7606)

**Comment by spena** [comment]
@mjsax I will follow-up the QTT tests in another PR. I will update and add some to test the "late" inputs. This PR only adds syntax. The grace period does not work yet. There's more code to add.

**Comment by spena** [comment]
@mjsax I updated the PR with the new proposed syntax (including KLIP-36). The new syntax is:
```
# Normal
JOIN <stream> WITHIN <window> [GRACE PERIOD <grace-period>]

# Before/After
JOIN <stream> WITHIN (<before-window>, <after-window>) [GRACE PERIOD <grace-period>]
```

File: design-proposals/klip-36-grace-stream-stream-joins.md
Comment (spena) [reply design-proposals/klip-36-grace-stream-stream-joins.md:105]: I have another follow-up PR to update the docs. I can update the KLIP (as it is a proposal), but I prefer to update the docs once the PR that enables the fix is merged.

**Comment by spena** [comment]
@mjsax I updated the PR to support the GRACE on before/after window syntax. The unit tests are still failing, but that's related to another issue 'cause ksql/master fails with the same.

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4
Comment (vcrfxia) [inline ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:231]: Are we planning to feature flag these changes in the event that not all the functionality makes it into the upcoming 0.20 release?

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java:906]: Is this change intentional? I assume it's fine if the tests below are passing, was just surprised to see it here.

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/WithinExpression.java
Comment (vcrfxia) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/WithinExpression.java:45]: nit: this is inconsistent with the constructor above. We should either update the constructor above so that all constructors call the one with all arguments, or we should update this constructor to parallel the one above by having this one call the one with `before` and `after` but without `location`.

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/tree/WithinExpressionTest.java
Comment (vcrfxia) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/tree/WithinExpressionTest.java:49]: nit:
```suggestion
  public void shouldDisplayCorrectStringWithGracePeriod() {
```

**Review by vcrfxia — APPROVED** [review]
Thanks @spena -- LGTM with comments inline.

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4
Comment (mjsax) [reply ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:231]: ksqlDB already has a feature flag to disable the stream-stream join fix -- I think it makes sense to use it to fail if `GRACE` is used and the feature was disabled. Do you want to get this pulled into this PR or a later PR?

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/WithinExpression.java
Comment (mjsax) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/WithinExpression.java:147]: This is actually "incorrect". We should apply GRACE PERIOD only via the new `JoinWindows#ofSizeAndGrace()` and never call `grace()` directly. -- Of course, we first need to get the AK PR merged. 

Just wanted to point it out, that we need to change this in a follow up PR. (Wondering though, we we need to apply GRACE in this PR at all? Could we not also just ignore it when building the plan until we can do it properly? Or do we need it for some testing?)

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java
Comment (mjsax) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:947]: To follow up to @vcrfxia comment -- should we add a test that verifies that we fail if `GRACE` is specified but if the feature flag disables the fix? (not sure if a unit test or QTT test should cover it?)

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java
Comment (mjsax) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:1009]: as above.

**Review by mjsax — APPROVED** [review]
Feel free to merge after `master` is fixed.

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4
Comment (spena) [reply ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:231]: I'll pull it into another PR

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java
Comment (spena) [reply ksqldb-engine/src/main/java/io/confluent/ksql/engine/rewrite/QueryAnonymizer.java:906]: Yes. The previous string builders used to append the empty character at the end. To avoid duplicating the effort, I just moved that empty char to the end of the method.

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/WithinExpression.java
Comment (spena) [reply ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/WithinExpression.java:147]: There's another jira is planned to call the new API, so a follow-up PR will include it once the AK PR is merged. Another thing is that this part of the code, which creates the `JoinWindows`, does not actually do anything with grace in the Kafka streams yet. Later on, another part of the code just gets the after/before windows from this JoinWindows, but not the grace period yet. This is just a preparation for the WithinExpression.

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java
Comment (spena) [reply ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:947]: I'll add a test in a follow-up PR

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java
Comment (spena) [reply ksqldb-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:1009]: I'll add a test in a follow-up PR

**Pull request merged**
